src/hotspot/share/gc/g1/g1CardSetContainers.inline.hpp

```diff
@@ -72,19 +76,29 @@ inline G1AddCardResult G1CardSetInlinePtr::add(uint card_idx, uint bits_per_card
   }
 }
 
-inline bool G1CardSetInlinePtr::contains(uint card_idx, uint bits_per_card) {
-  uint num_elems = num_cards_in(_value);
+inline uint G1CardSetInlinePtr::find(uint card_idx, uint bits_per_card, uint start_at, uint end_at) {
+  assert(start_at < end_at, "Precondition!");
+
   uintptr_t const card_mask = (1 << bits_per_card) - 1;
+  uintptr_t value = ((uintptr_t)_value) >> card_pos_for(start_at, bits_per_card);
 
-  uintptr_t value = ((uintptr_t)_value) >> card_pos_for(0, bits_per_card);
   // Check if the card is already stored in the pointer.
-  for (uint cur_idx = 0; cur_idx < num_elems; cur_idx++) {
+  for (uint cur_idx = start_at; cur_idx < end_at; cur_idx++) {
     if ((value & card_mask) == card_idx) {
-      return true;
+      return cur_idx;
     }
     value >>= bits_per_card;
   }
-  return false;
+  return end_at;
+}
+
+inline bool G1CardSetInlinePtr::contains(uint card_idx, uint bits_per_card) {
```

Comment:
I think `contains` previously didn't use it's `card_idx` argument.  Is it needed by any callers?  Or was the previous non-use just a performance bug?

Reply:
D'oh!  Of course it's used.  It's the thing we're looking for.  I just couldn't see it right there in front of me.  Never mind.